### PR TITLE
[DRAFT] Adding`MagneticMACE` 

### PR DESCRIPTION
**Pull request by CheukHinHoJerry**
This PR implements `MagneticMACE` as abstract class with other spin-informed variant of magnetic MACE. Major changes are 

- adding spin-informed MagneticMACE and related blocks as extension
- necessary changes for reading non-collinear spin data
- an ase calculator for using MagneticMACE
- TODO: adding functionality of fitting to magnetic forces

This PR is still WIP because I am not sure how one would like to place the code and I am current at the stage of adding some tests , happy to make changes. 

Questions in my mind for @ilyes319 atm:
- Should the model itself be implemented as an extension?
- Should those extra interaction blocks and functionalities that I need be added in `extension.py` too? Or should they be placed where they are supposed to be outside `extension.py`?

### CODE REVIEW

**Comment by ilyes319** [comment]
Hey @CheukHinHoJerry thank you very much for that. Yes the model should be in the extensions.py, the new blocks themselves can be in blocks.py. For the random rotations, could please create a new file in data folder called augmentations.py where you put it? thank you

**Comment by CheukHinHoJerry** [comment]
No problem - I will make those changes and let you know. Thank you for the quick comment.

**Comment by CheukHinHoJerry** [comment]
@ilyes319 I turned on precommit checks but there are too many lines in `blocks.py`. I can disable the warning or move the blocks that I have to other files. Any suggested fix?

**Comment by ilyes319** [comment]
@CheukHinHoJerry Yep you can disable that for that file!

**Comment by CheukHinHoJerry** [comment]
@ilyes319 Hi - after the very long silence I finally have this cleaned up. Would be good to have this merged before reforge? But happy to follow what you think would be good.

File: mace/modules/blocks.py [hunk 1]
```diff
@@ -1367,6 +1490,346 @@ def forward(
...
+        )
+        # transforming from radial l channels to magnetic l channels
+        self.conv_tp_weights_magmom = nn.FullyConnectedNet(
+            [
+                input_dim + magmom_input_dim,
+            ]
+            + [
+                self.magmom_conv_tp.weight_numel,
+            ]
+        )
```
[the remark this answers is not in the document]
Comment (CheukHinHoJerry) [reply mace/modules/blocks.py:1590]: I think this is irrelevant because evaluation tests passes.

File: mace/tools/arg_parser.py [hunk 2]
```diff
@@ -783,6 +806,20 @@ def build_default_arg_parser() -> argparse.ArgumentParser:
         default=100.0,
         dest="swa_forces_weight",
     )
+    parser.add_argument(
+        "--magforces_weight",
+        help="weight of mag forces loss",
+        type=float,
+        default=100.0,
+    )
```
[the remark this answers is not in the document]
Comment (CheukHinHoJerry) [reply mace/tools/arg_parser.py:814]: Fixed

File: mace/calculators/mace.py [hunk 3]
```diff
@@ -843,3 +843,520 @@ def get_descriptors(self, atoms=None, invariants_only=True, num_layers=-1):
...
+
+        Args:
+            atoms (ase.Atoms): The atomic structure to check.
+            tol (float): Tolerance for detecting changes.
+
+        Returns:
+            list: A list of changes detected in the system.
+        """
+        state = super().check_state(atoms, tol=tol)
+        if (not state) and (self.atoms.info != atoms.info):
```
[the remark this answers is not in the document]
Comment (CheukHinHoJerry) [reply mace/calculators/mace.py]: Fixed

File: mace/data/augmentation.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    is_shuffle = not isinstance(
+        original_loader.sampler, torch.utils.data.SequentialSampler
+    )
+
+    # Create new DataLoader with dataloader in mace with same parameters
+    new_loader = tg_mace.dataloader.DataLoader(
+        transformed_dataset,
+        batch_size=original_loader.batch_size,
+        shuffle=is_shuffle,
+        num_workers=original_loader.num_workers,
```
[the remark this answers is not in the document]
Comment (CheukHinHoJerry) [reply mace/data/augmentation.py]: Fixed

File: mace/modules/extensions.py [hunk 5]
```diff
@@ -988,3 +1007,741 @@ def forward(
...
+                    )
+                )
+            else:
+                self.readouts.append(
+                    readout_cls(
+                        hidden_irreps,
+                        o3.Irreps(f"{len(heads)}x0e"),
+                        cueq_config,
+                        oeq_config,
+                    )
```
[the remark this answers is not in the document]
Comment (CheukHinHoJerry) [reply mace/modules/extensions.py:1351]: Fixed

File: mace/tools/train.py [hunk 6]
```diff
@@ -229,6 +237,13 @@ def train(
             train_sampler.set_epoch(epoch)
         if "ScheduleFree" in type(optimizer).__name__:
             optimizer.train()
+
+        # allow data augmentation on magnetic moment
+        if data_aug_magmom:
+            from mace.data.augmentation import create_random_rotation_loader
+
+            train_loader = create_random_rotation_loader(train_loader)
```
Comment (aacostadiaz) [inline mace/tools/train.py]: This is inside the epoch loop and reassigns `train_loader`, so each epoch wraps another `TransformedDataset` around the previous one. Should it move above the `while` loop so it only wraps once?

File: mace/modules/utils.py [hunk 7]
```diff
@@ -164,26 +164,137 @@ def compute_hessians_loop(
...
+        assert (
+            magmoms is not None
+        ), "Magnetic momenet must be inputed to get magnetic forces"
+        forces, virials, stress, mag_forces = compute_forces_virials_magforces(
+            energy=energy,
+            positions=positions,
+            displacement=displacement,
+            cell=cell,
+            magmoms=magmoms,
+            training=(training or compute_hessian),
```
Comment (aacostadiaz) [inline mace/modules/utils.py]:
```suggestion
            training=(training or compute_hessian or compute_edge_forces),
```

File: mace/modules/utils.py [hunk 8]
```diff
@@ -192,6 +303,18 @@ def get_outputs(
...
+        mag_forces = None
+    elif compute_force and compute_magforces:
+        assert (
+            magmoms is not None
+        ), "Magnetic momenet must be inputed to get magnetic forces"
+        forces, mag_forces = compute_forces_magforces(
+            energy=energy,
+            positions=positions,
+            magmoms=magmoms,
+            training=(training or compute_hessian),
```
Comment (aacostadiaz) [inline mace/modules/utils.py]:
```suggestion
            training=(training or compute_hessian or compute_edge_forces),
```

File: mace/modules/extensions.py [hunk 9]
```diff
@@ -988,3 +1007,746 @@ def forward(
...
+        )
+        magmom_node_attrs = self.mag_solid_harmoics(data["magmom"])
+
+        #
+        magmom_node_feats = self.mag_radial_embedding(
+            magmom_lenghts_trans
+        )  # (n_atoms, n_basis)
+
+        # one body contribution radials, this is with constant shift so that it can be fitted
+        if self.use_magmom_one_body:
```
Comment (aacostadiaz) [inline mace/modules/extensions.py]: `use_magmom_one_body` is a plain bool, so TorchScript compiles both branches but `one_body_cheb_basis_with_const` only exists when the flag is on, so scripting the default model fails. `hasattr` folds at script time, which is why `pair_repulsion` uses it a couple of lines up.

```suggestion
        if hasattr(self, "one_body_cheb_basis_with_const"):
```

File: mace/tools/arg_parser.py [hunk 10]
```diff
@@ -1111,6 +1148,57 @@ def build_default_arg_parser() -> argparse.ArgumentParser:
...
+        default=8,
+    )
+    parser.add_argument(
+        "--use_magmom_one_body",
+        help="If true, use one body mangetic moment contribution in the model",
+        type=str2bool,
+        default=False,
+    )
+    parser.add_argument(
+        "--train_one_body_contribution",
```
Comment (aacostadiaz) [inline mace/tools/arg_parser.py:1190]: `--train_one_body_contribution` doesn't seem to be read anywhere

**Comment by CheukHinHoJerry** [comment]
I can't reproduce the pylint error locally... any suggested fix?

File: .pre-commit-config.yaml [hunk 11]
```diff
@@ -55,6 +55,6 @@ repos:
           '--disable=cell-var-from-loop',
           '--disable=duplicate-code',
           '--disable=use-dict-literal',
-          '--max-module-lines=1500',
+          '--max-module-lines=2000',
```
Comment (aacostadiaz) [inline .pre-commit-config.yaml]: Makes sense that `extensions.py` needs more room, but this lifts the limit for every module in the repo. `blocks.py` uses a per-file `# pylint: disable=too-many-lines instead`. Could extensions.py do the same

File: tests/extensions/magnetic/test_magmace.py [hunk 12]
```diff
@@ -0,0 +1,396 @@
+import argparse
```
Comment (aacostadiaz) [inline tests/extensions/magnetic/test_magmace.py:1]: The CI wiring for the new `magnetic` capability looks right. But the tests only check values are finite, so nothing pins down the symmetry of the new blocks. A rotation-equivariance test would be great: evaluate a config and a rotated copy (rotating positions and magmom by the same R), assert the energy is invariant and forces rotate with it. An inversion check (flip positions, hold spins)

**Review by aacostadiaz — COMMENTED** [review]
Thanks for this, really nice to see magnetic MACE taking shape!!

Don't worry about the red `lint` check, that one isn't on you. It's a pre-existing import cycle in `mace.data` that only surfaced because pylint isn't pinned and your new file shifted the module ordering. I'll fix it separately, so just ignore it here.

One heads-up: if you take the `--max-module-lines` suggestion below, add the per-file `# pylint: disable=too-many-lines` to `extensions.py` at the same time, otherwise lint just goes red on `C0302` instead.

**Comment by CheukHinHoJerry** [comment]
Thanks for the quick reply and confirming the pylint issue! I will take a look at those comment and fix it soon.

File: mace/tools/train.py (hunk 6 above)
Comment (CheukHinHoJerry) [reply mace/tools/train.py]: Fixed

File: mace/modules/utils.py (hunk 7 above)
Comment (CheukHinHoJerry) [reply mace/modules/utils.py]: Fixed

File: mace/modules/utils.py (hunk 8 above)
Comment (CheukHinHoJerry) [reply mace/modules/utils.py]: Fixed

File: mace/modules/extensions.py (hunk 9 above)
Comment (CheukHinHoJerry) [reply mace/modules/extensions.py]: Fixed.

File: mace/tools/arg_parser.py (hunk 10 above)
Comment (CheukHinHoJerry) [reply mace/tools/arg_parser.py:1190]: Thanks - there are some changes that I haven't copy from my dev branch. I have added some changes.

File: .pre-commit-config.yaml (hunk 11 above)
Comment (CheukHinHoJerry) [reply .pre-commit-config.yaml]: Applied the suggestions.

File: tests/extensions/magnetic/test_magmace.py (hunk 12 above)
Comment (CheukHinHoJerry) [reply tests/extensions/magnetic/test_magmace.py:1]: I have added global rotational and flip tests. Those I think are the expected symmetry for this model (mostly for efficiency concern so that they don't have exact symmetry.)

File: tests/extensions/magnetic/test_magmace.py [hunk 13]
```diff
@@ -0,0 +1,521 @@
...
+
+    # Z=99 isn't in z_table; should be ignored, not raise.
+    out = resolve_m_max(["{26: 1.8, 99: 1.0}"], [1, 6, 8, 26], default=1.0)
+    assert out == [1.0, 1.0, 1.0, 1.8]
+
+
+def test_resolve_m_max_numpy_atomic_numbers():
+    """atomic_numbers can be np.int64 (as it comes from z_table)."""
+    import numpy as np  # noqa: F401  (already imported at top of file)
+
```
Comment (theochemtheo) [inline tests/extensions/magnetic/test_magmace.py]:
```suggestion
```

File: mace/tools/finetuning_utils.py [hunk 14]
```diff
@@ -356,6 +386,255 @@ def load_foundations_elements(
...
+            num_species_foundations, -1
+        )[indices_weights, :]
+        .flatten()
+        .clone()
+        / (num_species_foundations / num_species) ** 0.5
+    )
+    if model.radial_embedding.bessel_fn.__class__.__name__ == "BesselBasis":
+        model.radial_embedding.bessel_fn.bessel_weights = torch.nn.Parameter(
+            model_foundations.radial_embedding.bessel_fn.bessel_weights.clone()
+        )
```
Comment (theochemtheo) [inline mace/tools/finetuning_utils.py]: does a similar thing need to be done for `GaussianBasis`  and `gaussian_weights`?

File: mace/tools/finetuning_utils.py [hunk 15]
```diff
@@ -356,6 +386,255 @@ def load_foundations_elements(
...
+                    ).weight.clone()
+                )
+
+        model.interactions[i].magmom_linear.weight = torch.nn.Parameter(
+            model_foundations.interactions[i].magmom_linear.weight.clone()
+        )
+        if model.interactions[i].__class__.__name__ in [
+            "MagneticRealAgnosticSpinOrbitCoupledDensityInteractionBlock",
+        ]:
+            # import pdb; pdb.set_trace()
```
Comment (theochemtheo) [inline mace/tools/finetuning_utils.py]: did you mean to keep the commented code here?

File: mace/tools/finetuning_utils.py [hunk 16]
```diff
@@ -356,6 +386,255 @@ def load_foundations_elements(
...
+                        .symmetric_contractions.contractions[j]
+                        .weights[k][indices_weights, :, :]
+                        .clone()
+                    )
+                )
+
+        model.products[i].conv_tp.weight = torch.nn.Parameter(
+            model_foundations.products[i].conv_tp.weight.clone()
+        )
+        # model.products[i].conv_tp_weights.weight = torch.nn.Parameter(
```
Comment (theochemtheo) [inline mace/tools/finetuning_utils.py]: should this also be retained?

File: requirements/magnetic.txt
Comment (AntObi) [inline requirements/magnetic.txt:1]: While the [LES](https://github.com/CheukHinHoJerry/mace/blob/abfedc2a461f069835b75c664fa77628c5e63919/requirements/les.txt) and [Polar](https://github.com/CheukHinHoJerry/mace/blob/abfedc2a461f069835b75c664fa77628c5e63919/requirements/polar.txt) require separate `requirements.txt` files, you could probably but the magnetic extras in the `setup.cfg` files given these dependencies exist on PyPI unlike the LES and Polar extras which point towards github repositories e.g.

In `setup.cfg`:
```
[options.extras_require]
...
magnetic = 
    sphericart-torch==1.0.9
    torch-geometric
```

**Comment by aacostadiaz** [comment]
@CheukHinHoJerry  thanks for addressing the comments. Could you merge the latest changes from `develop`? That should fix the lint issue.

**Comment by CheukHinHoJerry** [comment]
Thanks for the extra round of review. There are still some improvements to be done on how pre-train models are loaded but those can come later by a follow up PR I think.

File: mace/calculators/mace.py [hunk 17]
```diff
@@ -843,3 +844,532 @@ def get_descriptors(self, atoms=None, invariants_only=True, num_layers=-1):
...
+        self.results = {}
+        if self.model_type in ["MACE", "EnergyDipoleMACE"]:
+            self.results["energy"] = (
+                torch.mean(ret_tensors["energies"], dim=0).cpu().item()
+                * self.energy_units_to_eV
+            )
+            self.results["free_energy"] = self.results["energy"]
+            self.results["node_energy"] = (
+                torch.mean(ret_tensors["node_energy"], dim=0).cpu().numpy()
+            )
```
Comment (aacostadiaz) [inline mace/calculators/mace.py:1243]: Small one that I think slipped through last round: `node_energy` here misses the `energy_units_to_eV` factor that `energy` and `forces` get, so node energies come out in model units. The main MACECalculator does scale it.

```suggestion
            self.results["node_energy"] = (
                torch.mean(ret_tensors["node_energy"], dim=0).cpu().numpy()
                * self.energy_units_to_eV
            )
```

File: mace/data/atomic_data.py [hunk 18]
```diff
@@ -328,6 +351,20 @@ def from_config(
...
+            if config.properties.get("magmom") is not None
+            else None
+        )
+        magforces = (
+            torch.tensor(
+                config.properties.get("magforces"), dtype=torch.get_default_dtype()
+            )
+            if config.properties.get("magforces") is not None
+            else None
+        )
```
Comment (aacostadiaz) [inline mace/data/atomic_data.py:367]: `magmom`/`magforces` default to `None` where `forces`/`charges` default to zeros. Zeros would match how the other properties handle it.

**Review by aacostadiaz — COMMENTED** [review]
Thanks for addressing the comments. Two small things, then this looks good to merge from my side.

File: mace/calculators/mace.py (hunk 17 above)
Comment (CheukHinHoJerry) [reply mace/calculators/mace.py:1243]: Fixed! Thanks for catching it.

File: mace/data/atomic_data.py (hunk 18 above)
Comment (CheukHinHoJerry) [reply mace/data/atomic_data.py:367]: Fixed - some changes in the loss are adapted too so that they are defined correctly.
